src/Hazelcast.Net/CP/ClusterCPGroups.cs

```diff
@@ -0,0 +1,131 @@
+        private readonly ClusterMembers _clusterMembers;
+        private int _disposed;
+        private MemberConnection _listenerConnection;
+        private Task _listenerRegisterTask;
+        private long _correlationId;
+
+        private CancellationTokenSource _cts;
+
+        public const string CPGroupsJsonField = "cp.leaders";
+        public const long InitialVersion = -1;
+
+        public ClusterCPGroups(ClusterState clusterState, ClusterMembers members)
+        {
+            _clusterState = clusterState;
+            _clusterMembers = members;
+            _logger = _clusterState.LoggerFactory.CreateLogger<ClusterCPGroups>();
+        }
+
+        public long Version => _version;
+        public int Count => _groups.Count;
+
+        public void SetGroups(long version, ICollection<CPGroupInfo> groups, IList<KeyValuePair<Guid, Guid>> cpToApUuids)
+        {
+            if (_version >= version)
+                return;
+
+            var oldCount = _groups.Count;
+
+            var newGroup = MapCPtoAPUuids(cpToApUuids, groups);
+
+            lock (_mutex)
+            {
+                // One more check in safe env.
+                if (_version >= version)
+                    return;
+
+                _version = version;
+
+                _groups.Clear();
+                foreach (var g in newGroup)
+                {
+                    _groups[g.Key] = g.Value;
+                }
+
+            }
+
+            _logger.IfDebug()?.LogDebug("CP groups updated to Version {Version}." +
+                                        " Old groups count: {OldCount}, new count: {NewCount}", Version, oldCount, groups.Count);
+        }
+        private Dictionary<CPGroupId, Guid> MapCPtoAPUuids(IList<KeyValuePair<Guid, Guid>> cpToApUuids, ICollection<CPGroupInfo> groupInfos)
+        {
+            var mapIds = new Dictionary<CPGroupId, Guid>();
+
+            var leaderIds = cpToApUuids.ToDictionary(x => x.Key, x => x.Value);
+
+            foreach (var group in groupInfos)
+            {
+                if (leaderIds.TryGetValue(group.Leader.Uuid, out var apUuid))
+                {
+                    mapIds[group.GroupId] = apUuid;
+                }
+            }
+
+            return mapIds;
+        }
+
+
+        public Guid GetLeaderMemberId(CPGroupId groupId) => _groups.TryGetValue(groupId, out var memberId) ? memberId : Guid.Empty;
+
+        public bool TryGetLeaderMemberId(CPGroupId groupId, out Guid group) => _groups.TryGetValue(groupId, out group);
+
+        /// <summary>
+        /// Overrides current leader mappings with the given ones.
+        /// Suitable for first time leader mappings.
+        /// </summary>
+        /// <param name="cpGroupLeaders">Mapped CP Group Ids and their leaders</param>
+        public void SetCPGroupIds(IDictionary<CPGroupId, Guid> cpGroupLeaders)
+        {
+            lock (_mutex)
+            {
+                if(cpGroupLeaders.Count == 0) return;
```

Comment:
Is there a scenario where you expect `Count == 0`? Leadership updates are only sent by group leaders, so they should always contain at least 1 group, but I'm curious if you found a scenario where this is possible.

Reply:
Actually, I used a empty list or object instead of null. You can consider it as a null or emptiness check.